src/lib/AudioUltra/Media/AudioDecoder.ts

```diff
@@ -0,0 +1,260 @@
+
+      instance.renew();
+
+      AudioDecoder.cache.set(src, instance);
+
+      return instance;
+    }
+
+    super();
+
+    // only allow one cached decoder at a time to prevent memory leaks
+    // and limit the memory usage of the browser
+    if (AudioDecoder.cache.size > 0) {
+      for (const decoder of AudioDecoder.cache.values()) {
+        decoder.destroy();
+        decoder.cleanupCache();
+      }
+    }
+
+    AudioDecoder.cache.set(this.src, this);
+  }
+
+  get channelCount() {
+    return this._channelCount;
+  }
+
+  get sampleRate() {
+    return this._sampleRate;
+  }
+
+  get duration() {
+    return this._duration;
+  }
+
+  get dataLength() {
+    if (this.chunks && !this._dataLength) {
+      this._dataLength = this.chunks?.reduce((a, b) => a + b.length, 0) ?? 0;
+    }
+    return this._dataLength;
+  }
+
+  get dataSize() {
+    if (this.chunks && !this._dataSize) {
+      this._dataSize = this.chunks?.reduce((a, b) => a + b.byteLength, 0) ?? 0;
+    }
+    return this._dataSize;
+  }
+
+  get sourceDecoded() {
+    return this.chunks !== undefined;
+  }
+
+  get sourceDecodeCancelled() {
+    return this.cancelled && this.decodeId === 0;
+  }
+
+  /**
+   * Cancel the decoding process.
+   * This will stop the generator and dispose the worker.
+   */
+  cancel() {
+    this.cancelled = true;
+    this.decodeId = 0;
+    this.disposeWorker();
+  }
+
+  renew() {
+    this.cancelled = false;
+  }
+
+  destroy() {
+    // Since this is a singleton, we don't want to destroy the instance but clear all active
+    // subscriptions and cancel any pending decoding work
+    super.removeAllListeners();
+    this.cancel();
+  }
+
+  cleanupResolvers() {
+    this.decoderResolve?.();
+    this.decoderResolve = undefined;
+    this.decodingResolve?.();
+    this.decodingResolve = undefined;
+  }
+
+  /**
+   * Cleanup the decoder cache if it is no longer in use.
+   */
+  cleanupCache() {
+    delete this.chunks;
+    this.cleanupResolvers();
+    AudioDecoder.cache.delete(this.src);
+  }
+
+  /**
+   * Total number of chunks to decode
+   */
+  getTotalChunks(duration: number) {
+    return Math.ceil(duration / DURATION_CHUNK_SIZE);
+  }
+
+  /**
+   * Decode the audio file in chunks to ensure the UI remains responsive.
+   */
+  async decode(arraybuffer: ArrayBuffer, options?: { multiChannel?: boolean }): Promise<void> {
+    // decoding is already in progress for this source, so wait for it to finish
+    // If the worker has cached data, we can skip the decode step
+    if (this.cancelled || this.sourceDecoded) return Promise.resolve();
+
+    info('decode:start', this.src);
+
+    // clear any previous decoding state
+    this.decodeId = Date.now();
+    this.decodingPromise =  new Promise(resolve => (this.decodingResolve = resolve as any));
+    this.decoderPromise = new Promise(resolve => (this.decoderResolve = resolve as any));
+
+    await getAudioDecoderWorker(DecodeAudioWasm, arraybuffer)
```

Comment:
I think it's better not to mix Promise and async ways in one place?
```
try {
  const worker = await getAudioDecoderWorker(DecodeAudioWasm, arraybuffer);
  <Promise body>
} catch() { <can we skip errors?> }
<finally>
```

Reply:
It requires a 2 stage wait, which we need to know outside of this which has resolved for both when to show the loader vs when to begin drawing. It is also used to track the promises that have already succeeded, avoiding any further wasted work.